### PR TITLE
⚠️ clusterctl: add the clusterctl config cluster command

### PR DESCRIPTION
**Pull request by fabriziopandini**
**What this PR does / why we need it**:
This PR adds the `clusterctl config cluster` command, thus completing the day 1 journey

**Which issue(s) this PR fixes** 
Rif #1729

/assign @vincepri 
/cc @detiber @ncdc

### CODE REVIEW

File: cmd/clusterctl/pkg/client/common.go [hunk 1]
```diff
@@ -67,9 +67,8 @@ func parseProviderName(provider string) (name string, version string, err error)
 	}
 
 	name = t[0]
-	errs := validation.IsDNS1123Label(name)
-	if len(errs) != 0 {
-		return "", "", errors.Errorf("invalid name value: %s", strings.Join(errs, "; "))
+	if err := validateDNS1123Label(name); err != nil {
+		return "", "", errors.Errorf("invalid provider name %q. Provider name should be in the form name[:version] and the name should be valid", provider)
```
Comment (ncdc) [inline cmd/clusterctl/pkg/client/common.go]: This ignores `err` and doesn't return its details?

File: cmd/clusterctl/cmd/config_cluster.go [hunk 2]
```diff
@@ -81,14 +85,46 @@ func init() {
...
+	template, err := c.GetClusterTemplate(fo)
+	if err != nil {
+		return err
+	}
+
+	yaml, err := template.Yaml()
+	if err != nil {
+		return err
+	}
+	yaml = append(yaml, []byte("\n")...)
```
Comment (ncdc) [inline cmd/clusterctl/cmd/config_cluster.go]:
```suggestion
	yaml = append(yaml, '\n')
```

File: cmd/clusterctl/pkg/client/config_test.go [hunk 3]
```diff
@@ -68,22 +69,332 @@ func Test_clusterctlClient_GetProvidersConfig(t *testing.T) {
...
-				t.Errorf("GetProvidersConfig() error = %v, wantErr %v", err, tt.wantErr)
+			if err != nil {
+				if !tt.wantErr {
+					t.Fatalf("error = %v, wantErr %v", err, tt.wantErr)
+				}
 				return
 			}
+			if tt.wantErr {
+				t.Fatal("error = nil, wantErr not nil")
+			}
```
Comment (ncdc) [inline cmd/clusterctl/pkg/client/config_test.go]: xref https://github.com/kubernetes-sigs/cluster-api/pull/2048/files#r365875399

File: cmd/clusterctl/pkg/client/config_test.go [hunk 4]
```diff
@@ -68,22 +69,332 @@ func Test_clusterctlClient_GetProvidersConfig(t *testing.T) {
...
+			err := c.templateOptionsToVariables(tt.args.options)
+			if err != nil {
+				if !tt.wantErr {
+					t.Fatalf("error = %v, wantErr %v", err, tt.wantErr)
+				}
+				return
+			}
+			if tt.wantErr {
+				t.Fatal("error = nil, wantErr not nil")
+			}
```
Comment (ncdc) [inline cmd/clusterctl/pkg/client/config_test.go]: xref https://github.com/kubernetes-sigs/cluster-api/pull/2048/files#r365875399

File: cmd/clusterctl/pkg/client/config_test.go [hunk 5]
```diff
@@ -68,22 +69,332 @@ func Test_clusterctlClient_GetProvidersConfig(t *testing.T) {
...
+			got, err := client.GetClusterTemplate(tt.args.options)
+			if err != nil {
+				if !tt.wantErr {
+					t.Fatalf("error = %v, wantErr %v", err, tt.wantErr)
+				}
+				return
+			}
+			if tt.wantErr {
+				t.Fatal("error = nil, wantErr not nil")
+			}
```
Comment (ncdc) [inline cmd/clusterctl/pkg/client/config_test.go]: xref https://github.com/kubernetes-sigs/cluster-api/pull/2048/files#r365875399

File: cmd/clusterctl/pkg/client/client.go [hunk 6]
```diff
@@ -33,13 +33,29 @@ type InitOptions struct {
 	Force                   bool
 }
 
+// GetClusterTemplateOptions carries the options supported by GetClusterTemplate
+type GetClusterTemplateOptions struct {
+	Kubeconfig               string
+	Provider                 string
```
[the remark this answers is not in the document]
Comment (fabriziopandini) [reply cmd/clusterctl/pkg/client/client.go]: kk!

File: cmd/clusterctl/pkg/client/common.go (hunk 1 above)
Comment (fabriziopandini) [reply cmd/clusterctl/pkg/client/common.go]: fixed by using `errors.Wrapf`

**Comment by fabriziopandini** [comment]
@ncdc many thanks for the feedback! everything should be addressed now

**Comment by ncdc** [comment]
/lgtm

@vincepri PTAL

**Comment by fabriziopandini** [comment]
rebased

**Comment by fabriziopandini** [comment]
@ncdc this one should be ready now

File: cmd/clusterctl/pkg/client/config_test.go [hunk 7]
```diff
@@ -162,3 +163,302 @@ func Test_clusterctlClient_GetProviderComponents(t *testing.T) {
...
+			},
+			wantErr: true,
+		},
+		{
+			name: "fails for invalid control plane machine count",
+			args: args{
+				options: GetClusterTemplateOptions{
+					ClusterName:              "foo",
+					TargetNamespace:          "bar",
+					KubernetesVersion:        "v1.2.3",
```
Comment (Klaven) [inline cmd/clusterctl/pkg/client/config_test.go]: can `ClusterName`, `TargetNemespace` or `KubernetesVersion` be empty, is that valid? Should we test whatever the desired behavior for that?

Comment (fabriziopandini) [reply cmd/clusterctl/pkg/client/config_test.go]: templateOptionsToVariables handles validation and the defaulting rules when empty is passed. Nevertheless, I'm planning o do a crash test on the UX as soon as we complete all the functionalities

**Pull request merged**